dist/aurelia-dependency-injection.d.ts

```diff
@@ -282,224 +306,242 @@ export declare function singleton(keyOrRegisterInChild?: any, registerInChild?:
+  registerResolver<TBase = any, TImpl = any>(key: StringOrNever<TBase, TImpl>, resolver: Resolver<TBase, TImpl>): Resolver<TBase, TImpl>;
+  /**
+  * Registers a type (constructor function) by inspecting its registration annotations. If none are found, then the default singleton registration is used.
+  * @param key The key that identifies the dependency at resolution time; usually a constructor function.
+  * @param fn The constructor function to use when the dependency needs to be instantiated. This defaults to the key value when fn is not supplied.
+  */
+  autoRegister<TBase = any, TImpl = TBase>(key: DependencyCtorKeyOrNever<TBase, TImpl>, fn?: SubCtorOrFunctor<TBase>): Resolver<TBase, TImpl>;
+  autoRegister<TBase = any, TImpl = any>(key: StringOrNever<TBase, TImpl>, fn?: SubCtorOrFunctor<TBase>): Resolver<TBase, TImpl>;
+  /**
+  * Registers an array of types (constructor functions) by inspecting their registration annotations. If none are found, then the default singleton registration is used.
+  * @param fns The constructor function to use when the dependency needs to be instantiated.
+  */
+  autoRegisterAll<TImpl = any>(fns: SubCtorOrFunctor<TImpl>[]): void;
+  /**
+  * Unregisters based on key.
+  * @param key The key that identifies the dependency at resolution time; usually a constructor function.
+  */
+  unregister<TBase = any, TImpl = TBase>(key: DependencyCtorKeyOrNever<TBase, TImpl>): void;
+  unregister<TBase = any, TImpl = any>(key: StringOrNever<TBase, TImpl>): void;
+  /**
+  * Inspects the container to determine if a particular key has been registred.
+  * @param key The key that identifies the dependency at resolution time; usually a constructor function.
+  * @param checkParent Indicates whether or not to check the parent container hierarchy.
+  * @return Returns true if the key has been registred; false otherwise.
+  */
+  hasResolver<TBase = any, TImpl = TBase>(key: DependencyCtorKeyOrNever<TBase, TImpl>, checkParent?: boolean): boolean;
+  hasResolver<TBase = any, TImpl = any>(key: StringOrNever<TBase, TImpl>, checkParent?: boolean): boolean;
+  /**
+  * Gets the resolver for the particular key, if it has been registered.
+  * @param key The key that identifies the dependency at resolution time; usually a constructor function.
+  * @return Returns the resolver, if registred, otherwise undefined.
+  */
+  getResolver<TBase = any, TImpl = TBase>(key: DependencyCtorKeyOrNever<TBase, TImpl>): Resolver<TBase, TImpl>;
+  getResolver<TBase = any, TImpl = any>(key: StringOrNever<TBase, TImpl>): Resolver<TBase, TImpl>;
+  /**
+  * Resolves a single instance based on the provided key.
+  * @param key The key that identifies the object to resolve.
+  * @return Returns the resolved instance.
+  */
+  get<TBase, TImpl = TBase>(key: DependencyCtorKeyOrNever<TBase, TImpl>): TImpl;
+  get<TBase, TImpl = any>(key: StringOrNever<TBase, TImpl>): TImpl;
```

Comment:
these 2 are the main ones, so if we could have some thing to make sure it doesn't break TS, we're good I think. And from the look of this, it should be inferred properly. But I'd ask if @DrSammyD feels it's ok to have it in this form, w/o any way to verify

Reply:
I'd say we should convert to (or at least add) typescript versions of the tests to verify. I have a branch which I did that to verify this change, but if anybody else ends up modifying this,  it would probably break something.

If you agree, I can pull this branch into my master for this commit. It just needs an adapter to run ts the jasmine tests as typescript.
https://github.com/DrSammyD/dependency-injection/tree/ts-tests/test

Reply:
@bigopon has set up some of our TS infrastructure in other repos. So, I'll defer to him on that. He may have some recommendations on it for the sake of cross-repo consistency.

Reply:
Looks like aurelia/validation repo has typescript tests. I can copy the pattern over. I'll update the pull request when I get home.

Reply:
... So. This got a whole lot more involved than I initially planned. But here it is.